From 957c6c01f517654bab23a18ded533bbe639bfd57 Mon Sep 17 00:00:00 2001
From: Fabian Vogt <fabian@ritter-vogt.de>
Date: Sat, 9 Sep 2023 16:21:26 +0200
Subject: [PATCH 3/3] Let themes specify the used version of Qt

metadata.desktop in the theme directory has a new key "QtVersion" which
defaults to 5. If it's not 5, sddm-greeter-qt${QtVersion} will be used to
show the theme.
---
 src/common/ThemeMetadata.cpp |  6 ++++++
 src/common/ThemeMetadata.h   |  1 +
 src/daemon/Greeter.cpp       | 20 +++++++++++++++++---
 src/daemon/Greeter.h         |  1 +
 4 files changed, 25 insertions(+), 3 deletions(-)

diff --git a/src/common/ThemeMetadata.cpp b/src/common/ThemeMetadata.cpp
index 3bae0ff..2cb3762 100644
--- a/src/common/ThemeMetadata.cpp
+++ b/src/common/ThemeMetadata.cpp
@@ -28,6 +28,7 @@ namespace SDDM {
         QString mainScript { QStringLiteral("Main.qml") };
         QString configFile;
         QString translationsDirectory { QStringLiteral(".") };
+        int qtVersion = 5;
     };
 
     ThemeMetadata::ThemeMetadata(const QString &path, QObject *parent) : QObject(parent), d(new ThemeMetadataPrivate()) {
@@ -50,11 +51,16 @@ namespace SDDM {
         return d->translationsDirectory;
     }
 
+    int ThemeMetadata::qtVersion() const {
+        return d->qtVersion;
+    }
+
     void ThemeMetadata::setTo(const QString &path) {
         QSettings settings(path, QSettings::IniFormat);
         // read values
         d->mainScript = settings.value(QStringLiteral("SddmGreeterTheme/MainScript"), QStringLiteral("Main.qml")).toString();
         d->configFile = settings.value(QStringLiteral("SddmGreeterTheme/ConfigFile"), QStringLiteral("theme.conf")).toString();
         d->translationsDirectory = settings.value(QStringLiteral("SddmGreeterTheme/TranslationsDirectory"), QStringLiteral(".")).toString();
+        d->qtVersion = settings.value(QStringLiteral("SddmGreeterTheme/QtVersion"), 5).toInt();
     }
 }
diff --git a/src/common/ThemeMetadata.h b/src/common/ThemeMetadata.h
index 0612337..fdb89a9 100644
--- a/src/common/ThemeMetadata.h
+++ b/src/common/ThemeMetadata.h
@@ -36,6 +36,7 @@ namespace SDDM {
         const QString &mainScript() const;
         const QString &configFile() const;
         const QString &translationsDirectory() const;
+        int qtVersion() const;
 
         void setTo(const QString &path);
 
diff --git a/src/daemon/Greeter.cpp b/src/daemon/Greeter.cpp
index c0437ae..4974bb5 100644
--- a/src/daemon/Greeter.cpp
+++ b/src/daemon/Greeter.cpp
@@ -80,11 +80,26 @@ namespace SDDM {
         m_displayServerCmd = cmd;
     }
 
+    QString Greeter::greeterPathForQt(int qtVersion)
+    {
+        const QString suffix = qtVersion == 5 ? QString() : QStringLiteral("-qt%1").arg(qtVersion);
+        return QStringLiteral(BIN_INSTALL_DIR "/sddm-greeter%1").arg(suffix);
+    }
+
     bool Greeter::start() {
         // check flag
         if (m_started)
             return false;
 
+        // If no theme is given, use the default theme of the default greeter version
+        const int themeQtVersion = m_themePath.isEmpty() ? (QT_VERSION >> 16) : m_metadata->qtVersion();
+        QString greeterPath = greeterPathForQt(themeQtVersion);
+        if (!QFileInfo(greeterPath).isExecutable()) {
+            qWarning() << "The theme at" << m_themePath << "requires missing" << greeterPath << ". Using fallback theme.";
+            setTheme(QString());
+            greeterPath = greeterPathForQt(QT_VERSION >> 16);
+        }
+
         // themes
         QString xcursorTheme = mainConfig.Theme.CursorTheme.get();
         if (m_themeConfig->contains(QLatin1String("cursorTheme")))
@@ -139,7 +154,7 @@ namespace SDDM {
                 m_process->setProcessEnvironment(env);
             }
             // Greeter command
-            m_process->start(QStringLiteral("%1/sddm-greeter").arg(QStringLiteral(BIN_INSTALL_DIR)), args);
+            m_process->start(greeterPath, args);
 
             //if we fail to start bail immediately, and don't block in waitForStarted
             if (m_process->state() == QProcess::NotRunning) {
@@ -173,8 +188,7 @@ namespace SDDM {
 
             // command
             QStringList cmd;
-            cmd << QStringLiteral("%1/sddm-greeter").arg(QStringLiteral(BIN_INSTALL_DIR))
-                << args;
+            cmd << greeterPath << args;
 
             // greeter environment
             QProcessEnvironment env;
diff --git a/src/daemon/Greeter.h b/src/daemon/Greeter.h
index d25eb07..8cc996d 100644
--- a/src/daemon/Greeter.h
+++ b/src/daemon/Greeter.h
@@ -79,6 +79,7 @@ namespace SDDM {
         QProcess *m_process { nullptr };
 
         static void insertEnvironmentList(QStringList names, QProcessEnvironment sourceEnv, QProcessEnvironment &targetEnv);
+        static QString greeterPathForQt(int qtVersion);
     };
 }
 
-- 
2.42.0

